Tensor Network Simulator

merav-aharoni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
This is a new pull request for the tensor network simulator. 


### Details and comments
It is currently identical to pull request #58 created by Elad Goldman a few days ago. The reason for creating a new one is so that I can take responsibility for this branch, instead of Elad, who is working part-time. So it is safe to remove his pull request.
This is a very preliminary version that is not yet fully connected to aer. Future development should include full integration into aer, cleaning known (and unknown) bugs, and further improvements to the algorithm.


### TODO list
- [x] Write some tests
- [x] Write an entry in the CHANGELOG

atilag commented:
Thanks @merav-aharoni !
I have just closed #58 .
As @chriseclectic said in a previous comment, integrating with the rest of our Aer framework is going to be needed in order to merge this into the code base.
I can help with the integration, of course... so let's do something, whenever you think the code is close to a `feature complete` state, where there's only a few minor bugs to fix, but all the main features are already there (so there should not be major changes in the structure of the code), I can start reviewing it and giving instructions on how to integrate it with `Aer`.
Does it work for you?

merav-aharoni commented:
Thanks @atilag, that is exactly what we had in mind here. I certainly appreciate your offer to help with the integration and I intend to take you up on this offer.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-aer?pullRequest=64) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-aer?pullRequest=64) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: eladgoldman<br/>:x: merav-aharoni<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-aer?pullRequest=64) it.</sub>

merav-aharoni commented:
Hi @chriseclectic , @atilag , this pull request is ready for review. The basic functionality is working, including support for snapshots of type expectation value and statevector. We have also made structural changes, so I think we are stable now. There is more functionality to be added. I will update the issue with exactly what is already supported and what is still needed.

yehuda-naveh commented:
@merav-aharoni can you fix the conflicts, and  then check the travis if it fails? We'd like to have this version merged

merav-aharoni commented:
Hi @atilag , one of the tests failed on Travis (Linux) 
ImportError: Failed to import test module: terra.backends.test_qasm_simulator
it seems like something is deprecated in this test. It is not related to our code. Can you please advise?

merav-aharoni commented:
Hi @atilag , @chriseclectic , our pull request is quite stable now, up to date with the master, and passed all integration tests. Can you find some time to review it?

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  // To add B as new rows axis == 0 (vertical)
+  // To add A as new cols axis == 1 (horizontal)
+  if (axis != 0 && axis!= 1) {
+    std::cerr << "MU::concatenate: axis must be 0 or 1" << std::endl;
```
We use exceptions for errors that cannot be recovered from. In this case, we just need:
```c++
throw std::invalid_argument("axis must be 0 or 1");
```

merav-aharoni replied:
Done. I changed the error message to "Utils::concatenate: axis must be 0 or 1". I noticed other functions use "MU" rather than "Utils". I am not sure why. Let me know if I should change this message. Same for the next few comments.

atilag replied:
As this is an exception, we will know where it comes from because the stack is usually embedded in the exception information, so I'd explicitily write: "Utils::concatenate or MU", with the message: "axis must be 0 or 1" is enough.

merav-aharoni replied:
When I intentionally insert an error, all I get is "qasm_simulator: simulation failed" without the stack, and did not see the error message.  When I ran in a different way, I got the python stack and the error message, but not the C++. How do you run so as to see the error stack? is there some option to show this?

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  // To add A as new cols axis == 1 (horizontal)
+  if (axis != 0 && axis!= 1) {
+    std::cerr << "MU::concatenate: axis must be 0 or 1" << std::endl;
+	exit(1);
```
... and of course, we do not `exit` from the process.
Just remove this line.

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  matrix<T> temp = A;
+  if(axis == 0) {
+	if(cols1 != cols2) {
+	  std::cerr << "MU::concatenate: axis = 0 but cols_A =! cols_B" << std::endl;
```
Sames as above:
```c++
throw std::invalid_argument("axis is 0, but the numbers of columns are not equal");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  if(axis == 0) {
+	if(cols1 != cols2) {
+	  std::cerr << "MU::concatenate: axis = 0 but cols_A =! cols_B" << std::endl;
+	  exit(1);
```
... and no exit.

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  }
+  else if(axis == 1) {
+    if(rows1 != rows2) {
+	  std::cerr << "MU::concatenate: axis = 1 but rows_A =! rows_B" << std::endl;
```
```c++
throw std::invalid_argument("axis is 1, but the number of rows are not equal");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  else if(axis == 1) {
+    if(rows1 != rows2) {
+	  std::cerr << "MU::concatenate: axis = 1 but rows_A =! rows_B" << std::endl;
+	  exit(1);
```
no exit

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  // To split by rows, axis == 0 (Only when A rows divide by 2)
+  // To split by cols, axis == 1 (Only when A cols divide by 2)
+  if (axis != 0 && axis != 1) {
+    std::cerr << "MU::split: axis must be 0 or 1" << std::endl;
```
```c++
throw std::invalid_argument("axis must be 0 or 1");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  // To split by cols, axis == 1 (Only when A cols divide by 2)
+  if (axis != 0 && axis != 1) {
+    std::cerr << "MU::split: axis must be 0 or 1" << std::endl;
+    exit(1);
```
no exit

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+	  std::cerr << "MU::split: can't matrix A split by rows"  << std::endl;
+	  exit(1);
+	}
+  B.resize(rows/2 , cols);
```
can you indent this line and the next line?

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  matrix<T> temp = A;
+  if(axis == 0) {
+	if (rows % 2 != 0) {
+	  std::cerr << "MU::split: can't matrix A split by rows"  << std::endl;
```
```c++
throw std::invalid_argument("can't split first matrix argument by rows");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  }
+  else if(axis == 1) {
+    if (cols % 2 != 0) {
+	  std::cerr << "MU::split: can't split matrix A by cols"  << std::endl;
```
```c++
throw std::invalid_argument("can't split first matrix argument by columns");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+	}
+	B.resize(rows, cols/2);
+	C.resize(rows, cols/2);
+	for (size_t i = 0; i < rows; i++)
```
To avoid misunderstanding, I'd put the braces in this `for` loop:
```
for (size_t i = 0; i < rows; i++) {
```

merav-aharoni replied:
I agree.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+  size_t rows1 = A.GetRows(), rows2 = B.GetRows(), cols1 = A.GetColumns(),
+         cols2 = B.GetColumns();
+  if(rows1 != rows2 || cols1 != cols2) {
+	  std::cerr << "MU::elementwise_multiplication: matrices have different sizes"  << std::endl;
```
```c++
throw std::invalid_argument("matrices have different sizes");
```

merav-aharoni replied:
Done.

atilag commented on src/framework/utils.hpp:
```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
+         cols2 = B.GetColumns();
+  if(rows1 != rows2 || cols1 != cols2) {
+	  std::cerr << "MU::elementwise_multiplication: matrices have different sizes"  << std::endl;
+	  exit(1);
```
no exit

merav-aharoni replied:
Done.

atilag commented on src/simulators/qasm/qasm_controller.hpp:
```diff
@@ -39,8 +39,9 @@ namespace Simulator {
  * - "optimize_noise_threshold" (int): Qubit threshold for running circuit
  *   optimizations passes for a noisy circuit [Default: 12].
  * 
+>>>>>>> master
```
These are the remains of a git conflict

merav-aharoni replied:
Done.

atilag commented on src/simulators/qasm/qasm_controller.hpp:
```diff
@@ -394,7 +414,7 @@ void QasmController::initialize_state(const Circuit &circ,
     state.initialize_qreg(circ.num_qubits);
   } else {
-    state.initialize_qreg(circ.num_qubits, initial_state);
+    //state.initialize_qreg(circ.num_qubits, initial_state);
```
What's the reason for removing this line?

merav-aharoni replied:
I temporarily commented this out because of a problem I had. Now I put it back in.

merav-aharoni replied:
I put back this line. It was a temporary removal.

atilag commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,375 @@
+/**
```
Can we rename this file to lower case? and write some documentation in the header explaining what it's purpose.

merav-aharoni replied:
@atilag ,thanks for excellent comments! I will address them all.

merav-aharoni replied:
Maybe better to name it matrix_product_state.hpp/cpp? (rather than mps.hpp/cpp).
Also for SVD.hpp/cpp, do you think I should change it to svd.hpp/cpp?

atilag replied:
`matrix_product_state.hpp` and `svd.hpp` are fine.  We are still very much a header only library (but this is something that we may change in the near future), so designing your data structures to be used as .hpp is ok for now.

merav-aharoni replied:
Updated file names to lowercase.

merav-aharoni replied:
Should I add the IBM copyright notice at the beginning of each file?

merav-aharoni replied:
I added some documentation about the general idea and algorithm. I also made some fixes in the indentation and in the style of the comments to be more compatible with the rest of Aer.

merav-aharoni commented on src/base/controller.hpp:
```diff
@@ -339,21 +338,20 @@ void Controller::set_parallelization_experiments(const std::vector<Circuit>& cir
 
 void Controller::set_parallelization_circuit(const Circuit& circ) {
 
+
```
removed.

merav-aharoni commented on src/base/controller.hpp:
```diff
@@ -484,6 +481,7 @@ json_t Controller::execute(const json_t &qobj_js) {
   try {
     qobj.load_qobj_from_json(qobj_js);
   }
+
```
removed.

merav-aharoni commented on qiskit/providers/aer/aerjob.py:
```diff
@@ -72,7 +72,7 @@ def submit(self):
             raise JobError("We have already submitted the job!")
 
-        validate_qobj_against_schema(self._qobj)
+        #validate_qobj_against_schema(self._qobj)
```
I re-enabled the validation. This was a temporary hack that shouldn't have been committed.

atilag commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+			apply_swap(i,i-1);
+}
+
+cmatrix_t MPS::Density_matrix(const reg_t &qubits) const
```
Rename `Density_matrix` to `density_matrix`

merav-aharoni replied:
done.

atilag commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+  // ***** Assuming ascending sorted qubits register *****
+  vector<uint_t> internalIndexes;
+  for (uint_t index : qubits)
+    internalIndexes.push_back((uint_t)index);
```
Do we really need this casting?
If so, we don't use C-style castings, so:
```c++
internalIndexes.push_back(static_cast<uint_t>(index));
```

merav-aharoni replied:
Casting was not necessary. Removed it.

atilag commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+  MPS_Tensor psi = temp_TN.state_vec(new_indexes.front(), new_indexes.back());
+  uint_t size = psi.get_dim();
+  cmatrix_t rho(size,size);
+  for(uint_t i = 0; i < size; i++) {
```
This seems to be a good place for parallelizing.
Can you add: 
```c++
    #ifdef _WIN32
      #pragma omp for
    #else
      #pragma omp for collapse(2)
    #endif
    for(int_t i = 0; i < size; i++) {
 ```
Remember to change the `uint_t` from the `for(uint_t ...)` to `int_t` (due to the lack of proper support for OMP on Windows).

merav-aharoni replied:
Done.

atilag commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+  return real(res);
+}
+
+void MPS::printTN()
```
As I have commented above, overloading the << operator is considered a good practice for gathering debugging information from a class. It will return a `std::ostream` that you can use later within a `std::cout` for example.

merav-aharoni replied:
done.

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+vector<uint_t> calc_new_indexes(vector<uint_t> indexes);
+
+template <class T>
+void myswap(T &a, T &b){
```
Good idea!

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+  complex_t beta = 0.0f;
+  for(uint_t i = 0; i < num_qubits_; i++)
+      q_reg_.push_back(MPS_Tensor(alpha,beta));
+  for(uint_t i = 0; i < num_qubits_-1; i++)
```
United the loops.

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+	return new_indexes;
+}
+
+void MPS::initialize(uint_t num_qubits)
```
In principle, I agree with this comment as well as the next. However, I tried to be compatible with the other simulation methods. They all implement an empty constructor, and then they call initialize_qreg(), which calls qreg_.initialize(). Since initialize_qreg() is abstract in base class State, I must implement it. What do you think is best?

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+	q_reg_[index_B].mul_Gamma_by_right_Lambda(right_lambda);
+	MPS_Tensor temp = MPS_Tensor::contract(q_reg_[index_A],lambda_reg_[index_A], q_reg_[index_B]);
+
+	if(DEBUG) temp.print();
```
how do you compile the debug version?

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+    default:
+	  throw std::invalid_argument("illegal gate for apply_2_qubit_gate"); 
+	}
+	if(DEBUG) temp.print();
```
Done.

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+
+	if(DEBUG) temp.print();
+	temp.apply_swap();
+	if(DEBUG) temp.print();
```
done.

merav-aharoni commented on src/simulators/tensor_network/MPS.cpp:
```diff
@@ -0,0 +1,398 @@
+	  temp.apply_matrix(mat);
+	  break;
+
+    default:
```
done.

atilag commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
+{
+  if (data_.size() != 4)
+    {
+      cout << "ERROR: The tensor doesn't represent 2 qubits" << '\n';
```
Remove `cout`

merav-aharoni replied:
Removed this error - it was for our development phase.

atilag commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
+  if (data_.size() != 4)
+    {
+      cout << "ERROR: The tensor doesn't represent 2 qubits" << '\n';
+      assert(false);
```
```c++
throw std::logic_error("The tensor doesn't represent 2 qubits");
```

merav-aharoni replied:
Removed this error - it was for our development phase.

atilag commented on src/simulators/tensor_network/svd.cpp:
```diff
@@ -478,48 +474,9 @@ status csvd(cmatrix_t &A, cmatrix_t &U,rvector_t &S,cmatrix_t &V)
-//		cout << U*diag_S*(AER::Utils::dagger(V));
-//		cout << temp;
-//		assert(false);
+	  cout << "error: wrong SVD calc: A != USV*" << endl;
```
I think we need to put this into `#ifdef DEBUG` directives, right?
Is this a real error? Should we throw an exception here?... otherwise, I would change "error" with something different: "warning" ?

merav-aharoni replied:
This is a real error. If we get this error, the algorithm will not continue correctly., so it is better to exit. I changed to a "throw". Let me know if you think this should be handled differently.

atilag commented on src/simulators/tensor_network/svd.cpp:
```diff
@@ -530,30 +487,34 @@ status csvd(cmatrix_t &A, cmatrix_t &U,rvector_t &S,cmatrix_t &V)
+      current_status = csvd(A, U, S, V);
+    }
+  if(times == NUM_SVD_TRIES) {
+    std::cerr << "SVD failed" << std::endl;
```
We don't use standard output/error to write "logs" or "traces". I'd enclose this message with `#ifdef DEBUG/#endif` directives.

merav-aharoni replied:
This is also a real error. If we get this error, the algorithm will not continue correctly., so it is better to exit. I changed to a "throw". Let me know if you think this should be handled differently.

atilag commented on src/simulators/tensor_network/tensor_network_state.hpp:
```diff
@@ -336,16 +345,16 @@ void State::initialize_omp() {
-    cout << "initialize_creg not supported yet" <<endl;
-  }
+void State::initialize_creg(uint_t num_memory, uint_t num_registers) {
+  cout << "initialize_creg not supported yet" <<endl;
```
Enclose this with `#ifdef DEBUG` or just leave a comment and lave this method empty.
If enclosing, don't need to append `<<endl` in the end, just write:
```c++
#ifdef DEBUG
cout << "tensor_network_state::initialize_creg() not supported yet\n";
#endif
```

merav-aharoni replied:
enclosed in #ifdef DEBUG

atilag commented:
There are still two important things to complete before merging:
1. Write some tests
2. Write an entry in the CHANGELOG.md file

merav-aharoni commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
+  if (data_.size() != 2)
+    {
+      cout << "ERROR: The tensor doesn't represent one qubit" << '\n';
+      assert(false);
```
Removed this error - it was for our development phase.

merav-aharoni commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
-      data_[i](a1,a2) = data[i];
+    if (data_.size() != 2)
+      {
+	cout << "ERROR: The tensor doesn't represent one qubit" << '\n';
```
Removed this error - it was for our development phase.

merav-aharoni commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
+{
+  if (data_.size() != 2)
+    {
+      cout << "ERROR: The tensor doesn't represent one qubit" << '\n';
```
Removed this error - it was for our development phase.

merav-aharoni commented:
I added the test file: test/terra/backends/test_qasm_tensor_network.py. It is similar to test_qasm_extended_stabilizer_simulator.py. The only difference is that I list all the tests and run them in a loop.
I added an entry in CHANGELOG.md

merav-aharoni commented on src/simulators/tensor_network/matrix_product_state_tensor.hpp:
```diff
@@ -124,211 +135,197 @@ class MPS_Tensor
+{
+  if (data_.size() != 2)
+    {
+      cout << "ERROR: The tensor doesn't represent one qubit" << '\n';
```
removed this error.

chriseclectic commented on src/framework/utils.hpp:
```diff
@@ -110,6 +110,16 @@ template <class T> matrix<T> partial_trace_b(const matrix<T> &rho, size_t dimB);
 template <class T> matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B);
 
+// Matrix concatenate and split.
+template <class T> matrix<T> concatenate (const matrix<T> &A, const matrix<T> &B, uint_t axis);
```
Can you add the documentation about axis = 0 or 1 here for how the concatenation works, same for the `split` function

merav-aharoni replied:
I moved the documentation from the implementation to the header and improved it.

merav-aharoni commented on src/simulators/tensor_network/matrix_product_state.cpp:
```diff
@@ -41,6 +65,22 @@ uint_t reverse_bits(uint_t num, uint_t len) {
+  #ifdef _WIN32
+     #pragma omp for
+  #else
+     #pragma omp for collapse(1)
```
I fixed it as you suggested. Please have a look.
